examples/tutorial_3_streamproto.md

```diff
@@ -0,0 +1,94 @@
+In the [previous tutorial](tutorial_2_customproto.md), we dug a little deeper in to how the ping protocol works by creating a custom ping protocol ourselves.
+
+This tutorial will focus on how to create a custom streaming protocol using libp2p.
+
+# Custom protocol in libp2p
+Let's create a `part3.nim`, and import our dependencies:
+```nim
+import bearssl
+import chronos
+import stew/byteutils
+
+import libp2p
+```
+This is similar to the second tutorial. Next, we'll declare our custom protocol.
+```nim
+const TestStreamCodec = "/test/stream-proto/1.0.0"
+
+type TestStreamProto = ref object of LPProtocol
+```
+
+Just as we did in the last tutorial, we've set a [protocol ID](https://docs.libp2p.io/concepts/protocols/#protocol-ids), and created a custom `LPProtocol`.
+
+As in the last tutorial, we are going to handle our client and server parts separately. The server will wait for an instruction from the client, then send a stream of data to the client and finish by closing the connection. The client will listen in a loop until the server has finished and closed the connection. Then the client will close its end of the connection as well.
+
+Lets start with the server part:
+```nim
+proc new(T: typedesc[TestStreamProto]): T =
+  # every incoming connections will in be handled in this closure
+  proc handle(conn: Connection, proto: string) {.async, gcsafe.} =
+    echo "[server] received from client: ", string.fromBytes(await conn.readLp(1024))
+
+    # send stream of data
+    for i in 0..10:
+      await conn.writeLp("Part " & $i)
+
+    # We must close the connections ourselves when we're done with it
+    await conn.close()
+
+  return T(codecs: @[TestStreamCodec], handler: handle)
+```
+Again, `handle`, will be called for each incoming peer asking for this protocol. In our simple example, the client sends `"please send data"` and once received, the server echos the message and sends a stream of data back. In this example, we are simply sending a series of strings as the stream of data, but this can be any data in a practical application.
+
+We can now create our client part:
+```nim
+proc streamData(p: TestStreamProto, conn: Connection) {.async.} =
+  await conn.writeLp("please send data")
+  try:
+    # Read loop
+    while true:
+      let
+        # readLp reads length prefixed bytes and returns a buffer without the prefix
+        strData = await conn.readLp(1024)
+        str = string.fromBytes(strData)
+      echo "[client] received from server: ", str
+
+  except LPStreamEOFError:
+    # Once the remote peer has closed disconnected (closed the connection),
+    # we should also close the connection from this end as well.
+    # The only way to tell that a connection has been closed on the remote
```

Comment:
You can also use conn.closed() or conn.atEof() (which have different meanings because the connection can be half closed)

Reply:
I had tried both of these actually (and retried again since you mentioned it), and at no point inside the while loop is `conn.closed` or `conn.atEof` ever true. This is likely my lack of knowledge at how it is meant to work. What am I missing?

Reply:
Weird, need to look into that

Reply:
When using `readLp`, I think reason that `atEof` is not true is because `atEof` has already been checked and will not continue to be true on subsequent reads.

Reply:
Well, that's error prone.. Don't see how a stream atEof could not be atEof later
@dryajov we agree that is something that needs to be fixed?

Reply:
`readLp` and `atEof` weren't designed to be used together. The behavior of `readLp` is such that it will always await for all the data from the length prefixed message to be read before returning, which means that there is no way of checking `atEof` or `closed` when used in conjunction with it and other similar primitives such as `readExactly`, `readLine`, `readUntil`, etc... This is why an exception is raised to signal `EOF`. `atEof` and `closed` should however be true once the exception has been triggered. In general `atEof` is useful with `readOnce`, which optimistically reads up to the size of the passed buffer otherwise it returns as much as it was able to read. So, the behavior here is correct.

Reply:
Ok, either I or @emizzle misunderstood something:
```nim
  while not stream.atEof:
    try:
      let buf = await stream.readLp(1024)
      echo "[client] received from server: ", string.fromBytes(buf)
    except LPStreamEOFError:
      echo "finished: ", stream.atEof
  await stream.close()
```
Works fine (and echoes `finished: true`), so you _can_ use `atEof` with lp procs, as I expected

Though in this case you may as well just use the exception. But it's always nice to avoid a while true, imo

Reply:
Heh, this is a bit subtle - the reason it works is because you explicitly handle `LPStreamEOFError` and then check `atEof` after it fired, in this case it will be true, if however, you move the while into the `try`, you won't be able to use it as such.

Reply:
Ooohhh ok, I understand the misunderstanding.

You wanted to see if the stream was at eof before calling readLp, to avoid the exception raising. That's indeed not possible

And I though you were saying that once it throwed, isEof was still false, which I deemed wrong

So maybe a better wording would be something like "once the other end closes the connection, readLp will throw a [..] and stream.atEof will be true"